### PR TITLE
Optimized locking connection pool

### PR DESCRIPTION
**Pull request by TimonPost**
#102

### CODE REVIEW

File: src/net/connection/connection_pool.rs [hunk 1]
```diff
@@ -26,16 +26,23 @@ impl ConnectionPool {
...
-            .write()
+        let lock =  self.connections
+            .read()
             .map_err(|error| NetworkError::poisoned_connection_error(error.description()))?;
 
-        let connection = lock
-            .entry(*addr)
-            .or_insert_with(|| Arc::new(RwLock::new(VirtualConnection::new(*addr, &self.config))));
+        if lock.contains_key(addr) {
+            return Ok(lock.get(addr).unwrap().clone());
```
Comment (LucioFranco) [inline src/net/connection/connection_pool.rs]: Can you remove the return and the unwrap?

File: src/net/connection/connection_pool.rs [hunk 2]
```diff
@@ -26,16 +26,23 @@ impl ConnectionPool {
...
+        }else {
+            drop(lock);
```
Comment (LucioFranco) [inline src/net/connection/connection_pool.rs]: Good, so we don't create a deadlock 👍

File: src/net/connection/connection_pool.rs (hunk 1 above)
Comment (TimonPost) [reply src/net/connection/connection_pool.rs]: Strange though I already did that

**Comment by TimonPost** [comment]
bors r=LucioFranco

**Pull request merged**